features/netsocket/nsapi_types.h

```diff
@@ -48,6 +48,9 @@ enum nsapi_error {
     NSAPI_ERROR_DHCP_FAILURE  = -3010,     /*!< DHCP failed to complete successfully */
     NSAPI_ERROR_AUTH_FAILURE  = -3011,     /*!< connection to access point failed */
     NSAPI_ERROR_DEVICE_ERROR  = -3012,     /*!< failure interfacing with the network processor */
+    NSAPI_ERROR_IN_PROGRESS   = -3013,     /*!< operation (eg connect) in progress */
```

Comment:
It would be sort of be fine. BSD/POSIX would return ENOTCONN, so I'd expect stacks to give NO_CONNECTION to match.

Unfortunately in NSAPI, NSAPI will also give you NO_CONNECTION in place of ECONNFAILED, ECONNRESET or ETIMEDOUT. So you wouldn't be able to distinguish between a failed connection and one which just hadn't completed yet.

I guess you have to keep calling connect(), like this blocking loop.

And to make that work, you also have to make sure you do get a sensible error return from connect() after failure. What you wouldn't want is  for it to give up on a connection in the background, and then have the next connect call just start it again.

That's generally handled by one or both of - having sticky errors on sockets, so next socket call after an error returns the error code, and/or not permitting more than 1 connect() attempt on any socket.  Nanostack does the latter.